### PR TITLE
New function for mapping to project transformed point to face

### PR DESCRIPTION
**Pull request by thingishness**
Added a function to the mapping class that transforms a real to unit cell and then projects it to a dim-1 unit point on a face.

### CODE REVIEW

File: include/deal.II/fe/mapping.h [hunk 1]
```diff
@@ -276,6 +276,21 @@ class Mapping : public Subscriptor
...
+   * Transforms the point @p p on the real @p cell to the corresponding point
+   * on the unit cell, and then projects it to a dim-1  point on the face with
+   * the given face number @p face_no. Ideally the point @p p is near the face
+   * @ face_no, but any point in the cell can technically be projected.
+   *
+   * This function does not make physical sense when dim=1,
+   * so it throws an exception in this case.
+   */
+  virtual Point<dim-1>
+  transform_real_to_unit_projected_to_face (
```
Comment (tjhei) [inline include/deal.II/fe/mapping.h]: I would prefer a name like ```project_real_point_to_unit_point_on_face``` (project is better than transform)

**Comment by tjhei** [comment]
Looks good otherwise, though I would prefer seeing a test written for this function. Are you willing to do this? I am sure you can find somebody here to help with that. :-)

**Comment by thingishness** [comment]
I'm okay with the name change, and I've already been working on a unit test
along side it.  I'll see if I can finish the test tonight or tomorrow.

On Thu, Aug 6, 2015 at 6:19 PM, Timo Heister <notifications@github.com>
wrote:

> Looks good otherwise, though I would prefer seeing a test written for this
> function. Are you willing to do this? I am sure you can find somebody here
> to help with that. :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dealii/dealii/pull/1285#issuecomment-128524957>.
>



-- 
Jason Sheldon, M.S.
PhD Candidate, Department of Engineering Science and Mechanics
Graduate Research Assistant at the Pennsylvania State University

**Comment by bangerth** [comment]
@thingishness -- can you follow up with a separate pull request in which you note the addition of this function in `doc/news/changes.h`? Thanks in advance!

**Comment by bangerth** [comment]
@thingishness -- I think looking at the code that the function does not actually have to be `virtual`. Did you have a particular reason for doing so, e.g., because you thought that some derived class may want to implement it differently?

**Comment by thingishness** [comment]
@bangerth -- I suppose my only real reason was because it was in the base mapping class and there were a lot of derived classes, so I mirrored many of the other functions in mapping and used `virtual.  I there a negative to having it be `virtual`?

**Comment by bangerth** [comment]
There's no immediate drawback to it other than a reader expectation that virtual functions may and will be overridden in a derived class. Not marking it as virtual is essentially like saying "this is it -- no need to worry about what the implementation in some other class may look like".

**Pull request merged**